File: JAVA/level2/괄호_변환.java
class Solution {

    public String calculate(String p){
        if(p.equals(""))
            return "";

        int left = 0;
        int right = 0;
        int index = 0;
        boolean pass = true;

        for(int i=0; i<p.length(); i++){
            if(p.charAt(i) == '(')
                left++;
            if(p.charAt(i) == ')')
                right++;
            if(left < right)
                pass = false;
            if(left != 0 && left == right){
                index = i;
                break;
            }
        }

        if(pass == true)
            return p.substring(0, index+1) + calculate(p.substring(index+1));

        String str = p.substring(1, index);
        String temp = "";

        for(int i=0; i<str.length(); i++){
            if(str.charAt(i) == '(')
                temp += ")";
            if(str.charAt(i) == ')')
                temp += "(";
        }

        return ("("+ calculate(p.substring(index+1)) + ")" + temp);
    }

    public String solution(String p) {
        return calculate(p);
    }
}

File: JavaScript/level2/괄호_변환.js
function balance(p){
    let val = 0;
    
    for(let i=0; i<=p.length; i++){
        if(val === 0 && i != 0)
            return i;
    
        if(p[i] === "(")
            val++;
        else
            val--;
    }
    return -1;
}

function correct(p){
    let val = 0;
    
    for(let i=0; i<p.length; i++){
        if(val < 0)
            return false;
        
        if(p[i] === "(")
            val++;
        else
            val--;
    }
    return true;
}

function reverse(p){
    let answer = "";
    
    for(let i=0; i<p.length; i++){
        if(p[i] === "(")
            answer += ")"
        else
            answer += "("
    }
    return answer;
}

function calculate(p){
    if(p.length === 0)
        return "";
        
    let index = balance(p);
    let u = p.slice(0, index);
    let v = p.substring(index);
    
    if(correct(u))
        return (u + calculate(v));
    return ("(" + calculate(v) + ")" + reverse(u.slice(1, u.length-1)));
}

function solution(p) {
    return calculate(p);
}